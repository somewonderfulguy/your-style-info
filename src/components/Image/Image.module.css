@keyframes fadeIn {
  from {opacity: 0;}
  to {opacity: 1;}
}

.absolute {
  position: absolute;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
}

.figure {
  margin: 0;
}

.imageContainer {
  position: relative;
}

.img {
  max-width: 100%;
  animation: fadeIn .6s;
}

.imgAbsolute {
  composes: img;
  composes: absolute;
}

.aspectRatioOuter {
  width: 100%;
  margin: 0 auto;
  position: relative;
  overflow: hidden;
  transition: box-shadow 0.5s;
}

.aspectRatioOuterError {
  composes: aspectRatioOuter;
}

.aspectRatioOuterError:hover .aspectRatioInner {
  transform: scale(1.1);
}

.aspectRatioInner {
  width: 100%;
  overflow: hidden;
  position: relative;
  transition: transform 1s ease-in-out 0s;
}

.preloadPlaceholder {
  composes: absolute;

  background-size: cover;
  filter: blur(20px);
  transform: scale(1.05);
}

.caption {
  text-align: center;
  margin-top: 10px;
  color: rgba(0, 0, 0, 0.54);
  font-weight: 300;
  font-size: 15px;
}

.captionDark {
  composes: caption;
  color: rgba(255, 255, 255, 0.54);
}

.errorMessage {
  composes: absolute;
  background-color: rgba(0, 0, 0, .3);
}

.reloadImageBtn {
  composes: absolute;
  z-index: 5;

  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;

  color: white;
  width: 100%;
  height: 100%;

  cursor: pointer;
  outline: none;

  transition: transform 0.5s ease-in-out 0s;
}

.outline {
  composes: absolute;
  transition: all 0.2s ease;
}

.outlineDark {
  composes: outline;
}

.reloadImageBtn:focus + .outline {
  border: 1px rgba(33, 150, 243, 0.7) inset;
  box-shadow: inset 0 0 7px rgb(33, 150, 243);
}

.reloadImageBtn:focus + .outlineDark {
  border: 0.5px rgba(33, 150, 243, 0.7) inset;
  box-shadow: inset 0 0 3px rgb(33, 150, 243);
}

.reloadImageBtn:hover {
  transform: scale(1.05);
}

.reloadTitle {
  font-size: 20px;
  font-weight: 600;
  margin: 18px 0;
  position: absolute;
}

.reloadTitleSmaller {
  composes: reloadTitle;
  font-size: 17px;
}