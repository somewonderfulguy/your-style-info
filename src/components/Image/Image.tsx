import { MutableRefObject, useEffect, useReducer } from 'react'
import { animated, useTransition } from 'react-spring'

import { useImageLoadQuery } from '~api/imageQueries'
import { useIntersectionObserver, useResizeObserver } from '~shared/hooks'
import { useThemeState } from '~contexts/themeContext'
import { useIsDesktop } from '~contexts/screenDimensionsContext'
import classNames from '~shared/utils/classNames'

import pageStyles from '~components/Page/Page.module.css'
import styles from './Image.module.css'

type Props = {
  url: string
  alt: string
  lowresBase64?: string | null
  width?: number | string
  height: number
  caption?: string
  contentWidth?: 'content' | 'popup' | 'feature' | 'full'
}

// TODO: is it possible to use css aspect-ratio property?
export const getAspectRatio = (width: number, height: number) =>
  (height / width) * 100

// TODO: check mobiles
// TODO: different translations
// TODO: zoom to fullscreen (+ hide fullscreen on scroll)

const Image = ({
  url,
  alt,
  lowresBase64 = null,
  width = '100%',
  height,
  caption,
  contentWidth = 'content'
}: Props) => {
  const isDarkTheme = useThemeState()
  const isDesktop = useIsDesktop()
  const intersectionOffset = isDesktop ? 600 : 350
  const [bindInterObs, isIntersecting, disconnectIntersection] =
    useIntersectionObserver({
      rootMargin: `${intersectionOffset}px 0px ${intersectionOffset}px 0px`
    })
  const bindIntersectionObserver =
    bindInterObs as MutableRefObject<HTMLDivElement>

  const {
    isError,
    isSuccess,
    refetch: retry
  } = useImageLoadQuery(url, {
    enabled: Boolean(!!url && isIntersecting)
  })

  const [bindResObs, { width: imageWidth }] = useResizeObserver()
  const bindResizeObserver = bindResObs as MutableRefObject<HTMLDivElement>
  const isSmallerSize = imageWidth < 424

  type delayedType = {
    delayedShowTitle: boolean
    delayedShowSubtitle: boolean
  }
  const [{ delayedShowTitle, delayedShowSubtitle }, setAppear] = useReducer(
    (s: delayedType, a: Partial<delayedType>) => ({ ...s, ...a }),
    {
      delayedShowTitle: false,
      delayedShowSubtitle: false
    }
  )

  const errorTileTransitions = useTransition(isError, null, {
    from: { opacity: 0 },
    enter: { opacity: 1 },
    leave: { opacity: 0 }
  })

  const titleAppear = useTransition(delayedShowTitle, null, {
    from: {
      opacity: 0,
      top: `calc(50% - 19px)`
    },
    enter: {
      opacity: 1,
      top: `calc(50% - 39px)`
    },
    leave: { opacity: 0 },
    config: { duration: 400 }
  })

  const subtitleAppear = useTransition(delayedShowSubtitle, null, {
    from: {
      opacity: 0,
      top: `calc(50% + 30px)`
    },
    enter: {
      opacity: 1,
      top: `calc(50% + 20px)`
    },
    leave: { opacity: 0 },
    config: { duration: 400 }
  })

  useEffect(() => {
    if (isIntersecting) disconnectIntersection()
  }, [disconnectIntersection, isIntersecting])

  useEffect(() => {
    let timerTitle: NodeJS.Timeout | undefined
    let timerSubtitle: NodeJS.Timeout | undefined
    if (isError) {
      timerTitle = setTimeout(() => setAppear({ delayedShowTitle: true }), 400)
      timerSubtitle = setTimeout(
        () => setAppear({ delayedShowSubtitle: true }),
        700
      )
    } else {
      setAppear({
        delayedShowTitle: false,
        delayedShowSubtitle: false
      })
    }
    return () => {
      timerTitle && clearTimeout(timerTitle)
      timerSubtitle && clearTimeout(timerSubtitle)
    }
  }, [isError])

  return (
    <figure
      className={classNames(styles.figure, pageStyles[contentWidth])}
      ref={bindIntersectionObserver}
    >
      <div className={styles.imageContainer} ref={bindResizeObserver}>
        <div
          style={{ ...(contentWidth === 'content' && { maxWidth: width }) }}
          className={
            isError ? styles.aspectRatioOuterError : styles.aspectRatioOuter
          }
        >
          <div
            style={{
              paddingBottom: `${getAspectRatio(
                typeof width === 'string' ? +width.replace(/\D/g, '') : width,
                height
              )}%`
            }}
            className={styles.aspectRatioInner}
            aria-hidden
          >
            <div
              className={styles.preloadPlaceholder}
              style={{ backgroundImage: `url('${lowresBase64}')` }}
            />
          </div>
          {errorTileTransitions.map(
            ({ item, key, props }) =>
              item && (
                <animated.div
                  className={styles.errorMessage}
                  role="alert"
                  key={key}
                  style={props}
                >
                  <button
                    className={styles.reloadImageBtn}
                    onClick={() => retry()}
                    type="button"
                  >
                    {titleAppear.map(
                      ({ item, key, props }) =>
                        item && (
                          <animated.span
                            key={key}
                            className={
                              isSmallerSize
                                ? styles.reloadTitleSmaller
                                : styles.reloadTitle
                            }
                            style={props}
                          >
                            {/* TODO: translate */}
                            An error occured during image loading
                          </animated.span>
                        )
                    )}
                    {subtitleAppear.map(
                      ({ item, key, props }) =>
                        item && (
                          <animated.span
                            key={key}
                            style={props}
                            className={styles.reloadSubitle}
                          >
                            {/* TODO: translate */}
                            Click on this to try to load again.
                          </animated.span>
                        )
                    )}
                  </button>
                  <div
                    className={
                      isDarkTheme ? styles.outlineDark : styles.outline
                    }
                  />
                </animated.div>
              )
          )}
        </div>
        {isSuccess && (
          <img
            src={url}
            alt={alt}
            className={styles.imgAbsolute}
            style={{ ...(contentWidth !== 'content' && { width: '100%' }) }}
          />
        )}
      </div>
      {caption && (
        <figcaption
          className={isDarkTheme ? styles.captionDark : styles.caption}
        >
          {caption}
        </figcaption>
      )}
    </figure>
  )
}

export default Image
