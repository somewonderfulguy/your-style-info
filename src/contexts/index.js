export * from './headerHeightContext'
export * from './loadingContext'
export * from './localisationContext'
export * from './screenDimensionsContext'
export * from './themeContext'