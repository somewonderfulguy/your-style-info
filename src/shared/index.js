export * from './componentRenderer'
export * from './getComponent'
export * from './tests'