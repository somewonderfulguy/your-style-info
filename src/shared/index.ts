export * from './hooks'
export * from './tests'
export * from './utils'

export * from './types'