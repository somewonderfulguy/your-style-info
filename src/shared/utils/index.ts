export * from './debounce'
export * from './imgPreload'
export * from './isIpad'
export * from './throttle'
