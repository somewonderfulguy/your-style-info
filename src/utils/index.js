export * from './debounce'
export * from './imgPreload'