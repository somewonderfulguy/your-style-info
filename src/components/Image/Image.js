import React, {useEffect, useReducer} from 'react'
import {number, string} from 'prop-types'
import {animated, useTransition} from 'react-spring'

import {useImageLoad} from './hooks'
import {useTheme} from 'contexts'
import styles from './Image.module.css'

const propTypes = {
  url: string.isRequired,
  alt: string.isRequired,
  lowresBase64: string.isRequired,
  width: number.isRequired,
  height: number.isRequired,
  caption: string
}

const defaultProps = {
  lowresBase64: null,
  width: '100%'
}

export const getAspectRatio = (width, height) => height / width * 100

// TODO: check mobiles
// TODO: different translations
// TODO: zoom to fullscreen (+ hide fullscreen on scroll)

const Image = ({url, alt, lowresBase64, width, height, caption}) => {
  const {isRejected, isResolved, isPreviewVisible, retry} = useImageLoad(url)
  const {isDarkTheme} = useTheme()

  const [{delayedShowTitle, delayedShowSubtitle}, setAppear] = useReducer(
    (s, a) => ({...s, ...a}), {
      delayedShowTitle: false,
      delayedShowSubtitle: false
    }
  )

  const errorTileTransitions = useTransition(isRejected, null, {
    from: {opacity: 0},
    enter: {opacity: 1},
    leave: {opacity: 0}
  })

  const titleAppear = useTransition(delayedShowTitle, null, {
    from: {
      opacity: 0,
      top: `calc(50% - 19px)`
    },
    enter: {
      opacity: 1,
      top: `calc(50% - 39px)`
    },
    leave: {opacity: 0},
    config: {duration: 400}
  })

  const subtitleAppear = useTransition(delayedShowSubtitle, null, {
    from: {
      opacity: 0,
      top: `calc(50% + 30px)`
    },
    enter: {
      opacity: 1,
      top: `calc(50% + 20px)`
    },
    leave: {opacity: 0},
    config: {duration: 400}
  })

  useEffect(() => {
    let timerTitle, timerSubtitle
    if(isRejected) {
      timerTitle = setTimeout(() => setAppear({delayedShowTitle: true}), 400)
      timerSubtitle = setTimeout(() => setAppear({delayedShowSubtitle: true}), 700)
    } else {
      setAppear({
        delayedShowTitle: false,
        delayedShowSubtitle: false
      })
    }
    return () => {
      clearTimeout(timerTitle)
      clearTimeout(timerSubtitle)
    }
  }, [isRejected])

  return (
    <figure className={styles.figure}>
      <div className={styles.imageContainer}>
        {isPreviewVisible && (
          <div
            style={{maxWidth: width}}
            className={isRejected ? styles.aspectRatioOuterError : styles.aspectRatioOuter}
          >
            <div style={{paddingBottom: `${getAspectRatio(width, height)}%`}} className={styles.aspectRatioInner}>
              <div
                className={styles.preloadPlaceholder}
                style={{backgroundImage: `url('${lowresBase64}')`}}
              />
            </div>
            {errorTileTransitions.map(({item, key, props}) => (
              item && (
                <animated.div className={styles.errorMessage} role="alert" key={key} style={props}>
                  <button
                    className={styles.reloadImageBtn}
                    onClick={retry}
                    type="button"
                  >
                    {titleAppear.map(({item, key, props}) => (
                      item && (
                        <animated.span key={key} className={styles.reloadTitle} style={props}>
                          An error occured during image loading
                        </animated.span>
                      )
                    ))}
                    {subtitleAppear.map(({item, key, props}) => (
                      item && (
                        <animated.span key={key} style={{position: 'absolute', ...props}}>
                          Click on this to try to load again.
                        </animated.span>
                      )
                    ))}
                  </button>
                  <div className={isDarkTheme ? styles.outlineDark : styles.outline} />
                </animated.div>
              )
            ))
            }
          </div>
        )}
        {isResolved && (
          <img src={url} alt={alt} className={isPreviewVisible ? styles.imgAbsolute : styles.img} />
        )}
      </div>
      {caption && (
        <figcaption className={isDarkTheme ? styles.captionDark : styles.caption}>{caption}</figcaption>
      )}
    </figure>
  )
}

Image.propTypes = propTypes
Image.defaultProps = defaultProps

export default Image