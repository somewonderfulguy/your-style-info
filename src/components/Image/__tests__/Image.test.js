import React from 'react'
import {act, render, screen, wait, waitForElementToBeRemoved} from '@testing-library/react'
import user from '@testing-library/user-event'

import Image, {getAspectRatio} from '../Image'
import {imgPreloadPromise as mockImgPreloadPromise} from 'shared/utils'
import {ThemeProvider} from 'contexts'

jest.mock('shared/utils/imgPreload')

const FILE_PATH = 'path/nice-briefcase.jpg'
const CAPTION = 'A man in suit with briefcase and baguette'

const setup = props => {
  const utils = render(
    <Image
      url={FILE_PATH}
      alt={CAPTION}
      lowresBase64="base64string"
      width={1024}
      height={768}
      caption={CAPTION}
      {...props}
    />,
    {wrapper: ThemeProvider}
  )
  const getPreloadBlock = () => utils.container.querySelector('.preloadPlaceholder')
  return {
    getPreloadBlock,
    ...utils
  }
}

// TODO test focus on retry and zoom buttons

const mockPromise = Promise.resolve()
const mockCancel = jest.fn()
mockImgPreloadPromise.mockReturnValue([mockPromise, mockCancel])

afterEach(() => jest.clearAllMocks())

test('getAspectRatio fn calculates paddingBottom value correctly', () => {
  // TODO: replace with jest-in-case ??
  expect(getAspectRatio(16, 9)).toEqual(56.25)
  expect(+getAspectRatio(21, 9).toFixed(2)).toEqual(42.86)
})

test('image works as expected', async () => {
  const {getPreloadBlock, unmount} = setup()

  // preload block only, image loads and appears, preload block disappears
  expect(getPreloadBlock()).toBeInTheDocument()
  expect(screen.queryByRole('img')).not.toBeInTheDocument()
  await wait(() => expect(screen.getByRole('img')).toBeInTheDocument())
  await wait(() => expect(getPreloadBlock()).not.toBeInTheDocument())

  expect(mockImgPreloadPromise).toHaveBeenCalledWith(FILE_PATH)
  expect(mockImgPreloadPromise).toHaveBeenCalledTimes(1)

  // assert caption
  expect(screen.getByText(CAPTION)).toBeInTheDocument()

  // cancel image loading on unmount
  unmount()
  expect(mockCancel).toHaveBeenCalledTimes(1)
})

test('should not show figcaption if no caption passed as a prop', async () => {
  const {container} = setup({caption: undefined})
  const figcaption = container.querySelector('figcaption')
  expect(figcaption).not.toBeInTheDocument()

  await act(() => mockPromise)
})

test('"try again" block on failed image load works as expected', async () => {
  // FIXME: give it another chance after react-spring 9.0 released
  jest.spyOn(console, 'error').mockImplementation(() => {})

  const mockRejectPromise = Promise.reject(new Error('mock reject'))
  mockImgPreloadPromise.mockReturnValueOnce([mockRejectPromise, mockCancel])

  const {getPreloadBlock} = setup()

  // preloader, no image
  expect(getPreloadBlock()).toBeInTheDocument()
  expect(screen.queryByRole('img')).not.toBeInTheDocument()

  // error message appears
  const errorMsg = await screen.findByText(/an error occured/i)
  expect(errorMsg).toBeInTheDocument()

  // preloader placeholder is still in the document
  expect(getPreloadBlock()).toBeInTheDocument()

  // click on try again, error disappear and preload placeholder disappear, image appears
  act(() => user.click(screen.getByRole('button')))
  await waitForElementToBeRemoved(() => screen.getByText(/an error occured/i))
  await waitForElementToBeRemoved(getPreloadBlock)
  expect(screen.queryByRole('img')).toBeInTheDocument()

  await act(() => mockPromise)

  // assert mock
  expect(mockImgPreloadPromise).toHaveBeenCalledTimes(2)

  expect(console.error).toHaveBeenCalledTimes(1)
  jest.spyOn(console, 'error').mockRestore()
})