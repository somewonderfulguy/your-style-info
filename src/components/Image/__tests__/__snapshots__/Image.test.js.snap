// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <figure>
    <div
      class="imageContainer"
    >
      <div
        class="aspectRatioOuter"
        style="max-width: 100%;"
      >
        <div
          class="aspectRatioInner"
        >
          <div
            class="preloadPlaceholder"
            style="background-image: url(null);"
          />
        </div>
      </div>
    </div>
  </figure>
</DocumentFragment>
`;
